[ILLink] Add interface implementing type to OverrideInformation and use where it makes sense

jtschuster opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/98255

Adds the type that implements the interface to the `OverrideInformation` class to be able to properly handle cases where the base provides the interface method or the type relies on a default interface method. Adds the `InterfaceImplementor` class that is basically a tuple of `(TypeDefinition typeWithInterface, InterfaceImplementation ImplementedInterface)` to do this. Also cleans some unnecessary code in `OverrideInformation` and ensures that `OverrideInformation` will have an `InterfaceImplementor` if the base method is an interface method. 

This also reduces allocations of OverrideInformation in a hello world from ~35mb to ~1.2mb.

jtschuster committed:
Refactor OverrideInformation to include interface implementing type

jtschuster committed:
Add NotNullWhenAttribute for InterfaceImplementor

jtschuster committed:
Don't allow null interfaceImpl when base and override are interface methods in OverrideInformation

jtschuster committed:
Add InterfaceType to InterfaceImplementor

jtschuster committed:
Make new members internal for compat warnings

jtschuster committed:
Make InterfaceImplementor internal for API compat

jtschuster committed:
Make TypeMapInfo internal for API Compat

jtschuster committed:
Add InterfaceImplementor to suppressions.xml

jtschuster committed:
Make types public again

jtschuster committed:
Use correct InterfaceImplementor in FindAndAddDims

jtschuster committed:
InterfaceImplementor doesn't need to directly implement the interface

sbomer reviewed (commented):
~It's possible for a base to provide a static interface method, which seems like it might have broken the old logic. Could you add a testcase for this (using static interface methods, where the interface type isn't instantiated), and try to confirm that this change fixes it?~

edit: addressed by your comment at https://github.com/dotnet/runtime/pull/97487#discussion_r1492788849.

jtschuster commented on src/tools/illink/src/linker/Linker/TypeMapInfo.cs:
```diff
@@ -283,12 +300,12 @@ void AnnotateMethods (MethodDefinition @base, MethodDefinition @override, Interf
 		// most specific implementation of the given interface methods). ILLink needs to preserve
 		// all the implementations so that the proper exception can be thrown at runtime.
-		void FindAndAddDefaultInterfaceImplementations (TypeDefinition type, MethodDefinition interfaceMethod)
+		void FindAndAddDefaultInterfaceImplementations (TypeDefinition typeThatImplementsInterface, TypeDefinition typeThatMayHaveDimProvidingInterface, MethodDefinition interfaceMethod)
```
Yeah, good idea: https://github.com/dotnet/runtime/pull/98513

jtschuster commented on src/tools/illink/src/linker/Linker/OverrideInformation.cs:
```diff
@@ -3,71 +3,37 @@
+			Override = @override;
+			InterfaceImplementor = interfaceImplementor;
+			// Ensure we have an interface implementation if the base method is from an interface and the override method is on a class
+			Debug.Assert(@base.DeclaringType.IsInterface && interfaceImplementor != null
```
Yes, but getting that assert to work is blocked on https://github.com/dotnet/runtime/pull/98513 getting merged or added to this PR.

jtschuster committed:
PR Feedback

jtschuster committed:
Don't need to resolve TypeDefinitions

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into CheckImplementingType2

jtschuster committed:
Use correct InterfaceImpl

jtschuster committed:
Add argument to first call

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into CheckImplementingType2

jtschuster committed:
Look on base types and interfaces for interfaceImpl

jtschuster commented on src/tools/illink/src/linker/Linker/InterfaceImplementor.cs:
```diff
@@ -0,0 +1,68 @@
+		/// <summary>
+		/// The type of the interface that is implemented by <see cref="InterfaceImplementor.Implementor"/>
+		/// </summary>
+		public TypeDefinition InterfaceType { get; }
```
`InterfaceImplementation.InterfaceType` is a `TypeReference` and I thought it would be helpful to keep the `TypeDefinition` here. We could remove it and resolve it each time we need the definition, if you think that makes more sense.

jtschuster replied:
Yeah, that's a good idea. I'll assert in the constructor.

jtschuster commented on src/tools/illink/src/linker/Linker/InterfaceImplementor.cs:
```diff
@@ -0,0 +1,68 @@
+				}
+			}
+			var baseTypeRef = implementor.BaseType;
+			while (baseTypeRef is not null) {
```
In VisualBasic.Core.dll, VisualBasic.Collections implements IEnumerable through ICollection, but doesn't have the interfaceImpl, which causes failures in WASM test legs. I think to get a non-library test we need to fully implement the recursive interfaces issue, but I'll see if I can make a test case in library mode.

jtschuster committed:
Add test case and invert if

Revert "Look on base types and interfaces for interfaceImpl"

This reverts commit 81ebc93c265a7914b580e4fb62c0c77b078c9adb.

Reapply "Look on base types and interfaces for interfaceImpl"

This reverts commit 4320d2195719a68b51efa70fcced1fc5c7c9f6f2.

Add test case

Simplify test case

sbomer reviewed (approved):
LGTM aside from the last few comments, thanks!

I'm not sure how well this InterfaceImplementor concept will translate once you improve support for recursive interfaces, since the impl will no longer always be on the "Implementor" type. But looks good for the current support.

jtschuster committed:
PR feedback:

- Rename local
- Assert InterfaceType is resolved InterfaceImplementation.InterfaceType
